src/metabase/api/collection.clj

```diff
@@ -367,35 +367,60 @@
                    [:= :r.model_id :d.id]
                    [:core_user :u] [:= :u.id :r.user_id]]
        :where     [:and
+                   [:= :is_app_page page?]
                    [:= :collection_id (:id collection)]
                    [:= :archived (boolean archived?)]]}
       (hh/merge-where (pinned-state->clause pinned-state))))
 
+(defmethod collection-children-query :dashboard
+  [_ collection options]
+  (dashboard-query collection (assoc options :page? false)))
+
 (defmethod post-process-collection-children :dashboard
   [_ rows]
   (map #(dissoc %
                 :display :authority_level :moderated_status :icon :personal_owner_id :collection_preview
                 :dataset_query)
        rows))
 
-(defmethod collection-children-query :collection
-  [_ collection {:keys [archived? collection-namespace pinned-state]}]
+(defmethod collection-children-query :page
+  [_ collection options]
+  (dashboard-query collection (assoc options :page? true)))
+
+(defmethod post-process-collection-children :page
+  [_ rows]
+  (post-process-collection-children :dashboard rows))
+
+(defn- collection-query
+  [collection {:keys [app? archived? collection-namespace pinned-state]}]
   (-> (assoc (collection/effective-children-query
-               collection
-               [:= :archived archived?]
-               [:= :namespace (u/qualified-name collection-namespace)])
+              collection
+              [:= :archived archived?]
+              [:= :namespace (u/qualified-name collection-namespace)])
              ;; We get from the effective-children-query a normal set of columns selected:
              ;; want to make it fit the others to make UNION ALL work
              :select [:id
                       :name
                       :description
                       :entity_id
                       :personal_owner_id
-                      [(hx/literal "collection") :model]
-                      :authority_level])
+                      [(hx/literal (if app? "app" "collection")) :model]
+                      :authority_level
+                      :app_id]
+             ;; A simple left join would force us qualifying :id from
+             ;; collection and that doesn't work with effective-children-query.
+             ;; The sub-query makes sure that :app.id is only visible as :app_id.
+             :left-join [[{:select [[:id :app_id] :collection_id]
```

Comment:
Could this be done as a subselect on the select, if so I think it would be more clear? I suppose that makes the `:where` need to repeat it.

```clojure
{:select [
  .... 
  :authority_level
  [{:select [:app_id] :from [:app] :where [:= :app.collection_id :col.id]} :app_id]]}
```

Reply:
> I suppose that makes the `:where` need to repeat it.

@snoe  I think so, because the `:app_id` in the `:select` cannot be referenced in the `:where` clause. Do you think even repeating it there would be more readable?

Reply:
@metamben I think it's a tossup, so the `left-join` is fine.